homeassistant/components/withings/common.py

```diff
@@ -280,58 +655,393 @@ async def update_sleep_summary(self) -> SleepGetSummaryResponse:
+            final_exception = PlatformNotReady(last_exception)
+
+        await self._async_on_api_error(final_exception)
+        raise final_exception
+
+    async def async_webhook_data_updated(self, data_category: NotifyAppli) -> None:
+        """Handle scenario when data is updated from a webook."""
+        _LOGGER.info("Withings webhook triggered.")
+        print("BBBBBBB")
+        if data_category in {
+            NotifyAppli.WEIGHT,
+            NotifyAppli.CIRCULATORY,
+        }:
+            await self.async_update_measure()
+
+        if data_category in {NotifyAppli.SLEEP}:
+            await self.async_update_sleep_summary()
+
+        if data_category in {NotifyAppli.BED_IN, NotifyAppli.BED_OUT}:
+            await self.async_set_data(
+                WITHINGS_MEASUREMENTS_MAP[Measurement.IN_BED],
+                data_category == NotifyAppli.BED_IN,
+            )
+
+
+def get_attribute_unique_id(
+    attribute: WithingsAttribute, profile: str, user_id: int
+) -> str:
+    """Get a entity unique id for a user's attribute."""
+    measurement_slug = slugify(attribute.measurement.value)
+    return f"withings_{slugify(profile)}_{user_id}_" f"{measurement_slug}"
+
+
+async def async_get_entity_id(
+    hass: HomeAssistant, attribute: WithingsAttribute, profile: str, user_id: int
+) -> Optional[str]:
+    """Get an entity id for a user's attribute."""
+    entity_registry: EntityRegistry = await hass.helpers.entity_registry.async_get_registry()
+    unique_id = get_attribute_unique_id(attribute, profile, user_id)
+
+    entity_id = entity_registry.async_get_entity_id(
+        attribute.platform, const.DOMAIN, unique_id
     )
 
-    _LOGGER.debug("Creating withings data manager for profile: %s", profile)
-    return WithingsDataManager(hass, profile, api)
+    if entity_id is None:
+        _LOGGER.error("Cannot find entity id for unique_id: %s", unique_id)
+        return None
 
+    return entity_id
 
-def get_data_manager(
-    hass: HomeAssistant,
-    entry: ConfigEntry,
-    implementation: AbstractOAuth2Implementation,
+
+class BaseWithingsSensor(Entity):
+    """Base class for withings sensors."""
+
+    def __init__(
+        self, data_manager: WithingsDataManager, attribute: WithingsAttribute
+    ) -> None:
+        """Initialize the Withings sensor."""
+        self._data_manager = data_manager
+        self._attribute = attribute
+        self._state = STATE_UNKNOWN
+        self._profile = self._data_manager.profile
+        self._user_id = self._data_manager.user_id
+        self._unique_id = get_attribute_unique_id(
+            self._attribute, self._profile, self._user_id
+        )
+
+    @property
+    def should_poll(self) -> bool:
+        """Return False to indicate HA should not poll for changes."""
+        return False
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return f"Withings {self._attribute.measurement.value} {slugify(self._profile)}"
```

Comment:
Don't slugify in `name`. It's expensive call. Why not just include profile name? If you really need to slugify, do it once in the constructor.